File: birch.py
# Birch is a logging library for all your favorite languages
# by Constantin Volke (aka Wervice)

from datetime import UTC, datetime
from os import path

def get_utc_time():
    # Get current UTC time
    utc_time = datetime.now(UTC)

    # Format the time
    formatted_time = utc_time.strftime('%H:%M:%S:%f')[:-3]+utc_time.strftime(' %m/%d/%Y')

    return formatted_time

def color_text(text, color):
    colors = {
        'red': '\033[91m',
        'green': '\033[92m',
        'yellow': '\033[93m',
        'blue': '\033[94m',
        'cyan': '\033[96m'
    }
    end_color = '\033[0m'  # Reset color
    if color in colors:
        return f"{colors[color]}{text}{end_color}"
    else:
        return -1

def birch(level, message):

    try:
        birch_verbose
    except NameError:
        birch_verbose = True

    try:
        birch_path
    except NameError:
        raise Exception("Birch file path is not set.\nPlease define the variable birch_path")

    if path.exists(birch_path):
        with open(birch_path, "w") as log:
            log.write("Level\t\tTime\t\t\tMessage\n")

    if level == "verb" or level == "verbose" or level == 0 or level == "V":
        if birch_verbose:
            print(color_text(f"[VERBOSE] {get_utc_time()} | {message}", "blue"))
            with open(birch_path, "a") as birch_log:
                 birch_log.write(f"[VERBOSE]\t{get_utc_time()}\t{message}")
    elif level == "succ" or level == "success" or level == 1 or level == "S":
        print(color_text(f"[SUCCESS] {get_utc_time()} | {message}", "green"))
        with open(birch_path, "a") as birch_log:
                 birch_log.write(f"[SUCCESS]\t{get_utc_time()}\t{message}")
    elif level == "warn" or level == "warning" or level == 2 or level == "W":
        print(color_text(f"[WARNING] {get_utc_time()} | {message}", "yellow"))
        with open(birch_path, "a") as birch_log:
                 birch_log.write(f"[WARNING]\t{get_utc_time()}\t{message}")
    elif level == "erro" or level == "error" or level == 3 or level == "E":
        print(color_text(f"[ERROR] {get_utc_time()} | {message}", "red"))
        with open(birch_path, "a") as birch_log:
                 birch_log.write(f"[ERROR]\t\t{get_utc_time()}\t{message}")
    elif level == "fata" or level == "fatal" or level == 4 or level == "F":
        print(color_text(f"\x1b[1m[FATAL ERROR] {get_utc_time()} | {message}\x1b[0m", "red"))
        with open(birch_path, "a") as birch_log:
                 birch_log.write(f"[FATAL ERROR]\t{get_utc_time()}\t{message}")
    else:
        raise Exception("Birch does not know this type of logging level")

File: birch.js
// Spruce is a logging library for all your favorite languages
// by Constantin Volke (aka Wervice)

// Include this into your code by copying the lines bellow, by using eval() or using npm

const fs = require("fs");

function getUTCDateTime() {
	const now = new Date();

	// Get UTC components
	const utcHours = String(now.getUTCHours()).padStart(2, "0");
	const utcMinutes = String(now.getUTCMinutes()).padStart(2, "0");
	const utcSeconds = String(now.getUTCSeconds()).padStart(2, "0");
	const utcMilliseconds = String(now.getUTCMilliseconds()).padStart(3, "0");
	const utcDay = String(now.getUTCDate()).padStart(2, "0");
	const utcMonth = String(now.getUTCMonth() + 1).padStart(2, "0"); // Months are zero-based
	const utcYear = String(now.getUTCFullYear());

	// Construct the template
	const template = `${utcHours}:${utcMinutes}:${utcSeconds}:${utcMilliseconds} ${utcMonth}/${utcDay}/${utcYear}`;

	return template;
}

function colorText(text, color) {
	const colors = {
		dimmed: "\x1b[2m",
		red: "\x1b[31m",
		green: "\x1b[32m",
		yellow: "\x1b[33m",
		blue: "\x1b[34m",
		cyan: "\x1b[36m",
	};

	const reset = "\x1b[0m";

	if (colors[color]) {
		return colors[color] + text + reset;
	} else {
		return text; // Return text without color if color not found
	}
}

function birch(level, message) {
	if (typeof birch_path == "undefined") {
		throw new Error(
			"Birch file path is not set.\nPlease define the variable birch_path",
		);
	}

	if (typeof birch_verbose == "undefined") {
		birch_verbose = true;
	}

	if (!fs.existsSync(birch_path))	fs.writeFileSync(birch_path, "Level\t\tTime\t\t\tMessage\n");

	if (level == "verb" || level == "verbose" || level == 0 || level == "V") {
		if (birch_verbose == true) {
			console.log(
				colorText(`[VERBOSE] @ ${getUTCDateTime()} | ${message}`, "blue"),
			);
		fs.appendFileSync(
			birch_path,
			`[VERBOSE]\t${getUTCDateTime()}\t${message}\n`,
		);
		}
	} else if (
		level == "succ" ||
		level == "success" ||
		level == 1 ||
		level == "S"
	) {
		console.log(
			colorText(`[SUCCESS] @ ${getUTCDateTime()} | ${message}`, "green"),
		);
		fs.appendFileSync(
			birch_path,
			`[SUCCESS]\t${getUTCDateTime()}\t${message}\n`,
		);
	} else if (
		level == "warn" ||
		level == "warning" ||
		level == 2 ||
		level == "W"
	) {
		console.log(
			colorText(`[WARNING] @ ${getUTCDateTime()} | ${message}`, "yellow"),
		);
		fs.appendFileSync(
			birch_path,
			`[WARNING]\t${getUTCDateTime()}\t${message}\n`,
		);
	} else if (
		level == "erro" ||
		level == "error" ||
		level == 3 ||
		level == "E"
	) {
		console.log(colorText(`[ERROR] @ ${getUTCDateTime()} | ${message}`, "red"));
		fs.appendFileSync(
			birch_path,
			`[ERROR]\t\t${getUTCDateTime()}\t${message}\n`,
		);
	} else if (
		level == "fata" ||
		level == "fatal" ||
		level == 4 ||
		level == "F"
	) {
		console.log(
			colorText(
				`\x1b[1m[FATAL ERROR] @ ${getUTCDateTime()} | ${message}\x1b[0m`,
				"red",
			),
		);
		fs.appendFileSync(
			birch_path,
			`[FATAL ERROR]\t${getUTCDateTime()}\t${message}\n`,
		);
	} else {
		throw new Error("Birch does not know this type of logging level");
	}
}
